torch/csrc/jit/dynamic_dag.h

```diff
@@ -0,0 +1,388 @@
+  // the vertices that can reach producer.
+  reorder(deltaF_, deltaB_);
+}
+
+// Define the affected region for contractEdge(producer, consumer) as
+// { v s.t. ord(v) in [ord(producer), ord(consumer)] }.
+// These are the only vertices that can possibly be moved around
+// during edge contraction.
+//
+// contractEdge is O(|AR| * |out_edges(producer)|)
+template <typename T>
+bool DynamicDAG<T>::contractEdge(Vertex<T>* producer, Vertex<T>* consumer) {
+  JIT_ASSERT(producer != consumer);
+  if (contractionProducesCycle(producer, consumer)) {
+    return false;
+  }
+
+  removeEdge(producer, consumer);
+
+  // We can either merge producer into consumer or the other way around.
+  // I've chosen to merge producer into consumer for now but this is arbitrary.
+  consumer->rdata.insert(consumer->rdata.end(), producer->rdata.begin(), producer->rdata.end());
+
+  vertex_set<T> in_edges;
+  vertex_set<T> out_edges;
+  std::tie(in_edges, out_edges) = removeVertex(producer);
+
+  // Each of these are constant b/c ord(consumer) > ord(producer) > ord(parent)
+  // so the edge addition still preserves the existing topological order.
+  for (auto* parent : in_edges) {
+    addEdge(parent, consumer);
+  }
+
+  // NB: each addEdge call is linear in (ord(consumer) - ord(child)).
+  // This makes contractEdges O(|out_edges(producer)| * |AR|).
+  for (auto* child : out_edges) {
+    addEdge(consumer, child);
+  }
+  return true;
+}
+
+template <typename T>
+bool DynamicDAG<T>::contractionProducesCycle(Vertex<T>* producer, Vertex<T>* consumer) {
+  // Basically a modified dfsForward
+  vertex_list<T> stack;
+  vertex_list<T> visited;
+  stack.push_back(producer);
+
+  size_t upper_bound = consumer->ord;
+
+  while (!stack.empty()) {
+    auto* vertex = stack.back();
+    stack.pop_back();
+
+    if (vertex->visited) continue;
+    vertex->visited = true;
+    visited.push_back(vertex);
+
+    for (auto* next : vertex->out_edges) {
+      if (vertex != producer && next->ord == upper_bound) {
+        clearVisited(visited);
+        return true;
+      }
+      if (next->ord < upper_bound) {
+        stack.push_back(next);
+      }
+    }
+  }
+  clearVisited(visited);
+  return false;
+}
+
+
+// Performs a forward DFS.
+// Marks vertices as 'visited' and inserts them into visited.
+// Returns false immediately if upper_bound is reached.
+// Returns true otherwise.
+template <typename T>
+bool DynamicDAG<T>::dfsForward(Vertex<T>* producer, size_t upper_bound,
+    vertex_list<T>& visited) {
+  vertex_list<T> stack;
+  stack.push_back(producer);
+
+  while (!stack.empty()) {
+    auto* vertex = stack.back();
+    stack.pop_back();
+
+    if (vertex->visited) continue;
+    vertex->visited = true;
+    visited.push_back(vertex);
+
+    for (auto* next : vertex->out_edges) {
+      if (next->ord == upper_bound) {
+        return false;
+      }
+      if (next->ord < upper_bound) {
+        stack.push_back(next);
```

Comment:
It's faster if you move the `visited` check to here and avoid pushing those node altogether (otherwise, set it to true and push).